datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,858 @@
+        // Distribution and Ordering enforcement need to be applied bottom-up.
+        let target_partitions = config.target_partitions;
+        transform_up(plan, &{
+            |plan| Some(ensure_distribution_and_ordering(plan, target_partitions))
+        })
+    }
+
+    fn name(&self) -> &str {
+        "BasicEnforcement"
+    }
+}
+
+fn ensure_distribution_and_ordering(
+    plan: Arc<dyn crate::physical_plan::ExecutionPlan>,
+    target_partitions: usize,
+) -> Arc<dyn crate::physical_plan::ExecutionPlan> {
+    if plan.children().is_empty() {
+        return plan;
+    }
+    let required_input_distributions = plan.required_input_distribution();
+    let required_input_orderings = plan.required_input_ordering();
+    let children: Vec<Arc<dyn ExecutionPlan>> = plan.children();
+    assert_eq!(children.len(), required_input_distributions.len());
+    assert_eq!(children.len(), required_input_orderings.len());
+
+    // Add RepartitionExec to guarantee output partitioning
+    let children = children
+        .into_iter()
+        .zip(required_input_distributions.into_iter())
+        .map(|(child, required)| {
+            if child
+                .output_partitioning()
+                .satisfy(required.clone(), || child.equivalence_properties())
+            {
+                child
+            } else {
+                let new_child: Arc<dyn ExecutionPlan> = match required {
+                    Distribution::SinglePartition
+                        if child.output_partitioning().partition_count() > 1 =>
+                    {
+                        Arc::new(CoalescePartitionsExec::new(child.clone()))
+                    }
+                    _ => {
+                        let partition = required.create_partitioning(target_partitions);
+                        Arc::new(RepartitionExec::try_new(child, partition).unwrap())
+                    }
+                };
+                new_child
+            }
+        });
+
+    // Add SortExec to guarantee output ordering
+    let new_children: Vec<Arc<dyn ExecutionPlan>> = children
+        .zip(required_input_orderings.into_iter())
+        .map(|(child, required)| {
+            if ordering_satisfy(child.output_ordering(), required, || {
+                child.equivalence_properties()
+            }) {
+                child
+            } else {
+                let sort_expr = required.unwrap().to_vec();
+                if child.output_partitioning().partition_count() > 1 {
+                    Arc::new(SortExec::new_with_partitioning(
+                        sort_expr, child, true, None,
+                    ))
+                } else {
+                    Arc::new(SortExec::try_new(sort_expr, child, None).unwrap())
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    with_new_children_if_necessary(plan, new_children).unwrap()
+}
+
+/// DynamicEnforcement rule
+///
+///
+#[derive(Default)]
+pub struct DynamicEnforcement {}
+
+// TODO
```

Comment:
What is this supposed to do?

Reply:
> This is looking pretty neat @mingmwang but it is basically impossible for me to carefully review such a large PR. I wonder if you can break it down into pieces so we can get it in?
> 
> Possible split:
> 
> 1. Add `PhysicalExpr::children()` and `PhysicalExpr::new_with_children`
> 2. the code for iterating / manipulating physical exprs
> 3. The code for calculating equivalence properties (and tests)
> 4. The code in datafusion/core/src/physical_optimizer/enforcement.rs

Sure, I will split this PR to the 3 PRs:

1. PhysicalExpr related change: implement PartialEq<dyn Any> for PhysicalExpr, PhysicalExpr::children() and PhysicalExpr::new_with_children, the utils code for iterating / manipulating physical exprs.
2. Execution plan related changes: output_partitioning(), output_ordering(), required_input_distribution(), the code for calculating equivalence properties (and tests)
3. The code in datafusion/core/src/physical_optimizer/enforcement.rs and the tests.

Reply:
Thank. you very much